// vite-plugin-lib-inject-css/dist/index.js
// src/index.ts
import path from "path";
import color from "picocolors";
import MagicString from "magic-string";
var createPreserveModulesWarning = (optionPath) => "When `" + optionPath + "` is `true`, the association between chunk file and its css references will lose, so the style code injection will be skipped.";
function libInjectCss() {
  let skipInject = false;
  let resolvedConfig;
  return {
    name: "vite:lib-inject-css",
    apply: "build",
    enforce: "post",
    config({ build }) {
      for (const item of [build?.rollupOptions?.output].flat()) {
        if (item && typeof item.hoistTransitiveImports !== "boolean") {
          item.hoistTransitiveImports = false;
        }
      }
      return {
        build: {
          /**
           * Must enable css code split, otherwise there's only one `style.css` and `chunk.viteMetadata.importedCss` will be empty.
           * @see https://github.com/vitejs/vite/blob/HEAD/packages/vite/src/node/plugins/css.ts#L613
           */
          cssCodeSplit: true,
          /**
           * Must emit assets on SSR, otherwise there won't be any CSS files generated and the import statements
           * injected by this plugin will refer to an undefined module.
           * @see https://github.com/vitejs/vite/blob/HEAD/packages/vite/src/node/plugins/asset.ts#L213-L218
           */
          ssrEmitAssets: true
        }
      };
    },
    configResolved(config) {
      resolvedConfig = config;
    },
    options() {
      const { build, command } = resolvedConfig;
      const outputOptions = [build.rollupOptions.output].flat();
      const messages = [];
      if (!build.lib || command !== "build") {
        skipInject = true;
        messages.push(
          "Current is not in library mode or building process, skip code injection."
        );
      }
      if (outputOptions.some((v) => v?.preserveModules === true)) {
        skipInject = true;
        messages.push(
          createPreserveModulesWarning("rollupOptions.output.preserveModules")
        );
      }
      if (parseInt(this.meta.rollupVersion) < 4 && // @ts-ignore
      build.rollupOptions.preserveModules === true) {
        skipInject = true;
        messages.push(
          createPreserveModulesWarning("rollupOptions.preserveModules")
        );
      }
      if (build.ssr && build.ssrEmitAssets === false) {
        messages.push(
          "`config.build.ssrEmitAssets` is set to `true` by the plugin internally in library mode, but it seems to be `false` now. This may cause style code injection to fail on SSR, please check the configuration to prevent this option from being modified."
        );
      }
      messages.forEach(
        (msg) => console.log(
          `
${color.cyan("[vite:lib-inject-css]")} ${color.yellow(msg)}
`
        )
      );
    },
    renderChunk(code, chunk) {
      if (skipInject || !chunk.viteMetadata)
        return;
      const { importedCss } = chunk.viteMetadata;
      if (!importedCss.size)
        return;
      const ms = new MagicString(code);
      for (const cssFileName of importedCss) {
        let cssFilePath = path.relative(path.dirname(chunk.fileName), cssFileName).replaceAll(/[\\/]+/g, "/");
        cssFilePath = cssFilePath.startsWith(".") ? cssFilePath : `./${cssFilePath}`;
        ms.prepend(`import '${cssFilePath}';
`);
      }
      return {
        code: ms.toString(),
        map: ms.generateMap()
      };
    }
  };
}
export {
  libInjectCss
};
